src/android/java/io/jxcore/node/BtToRequestSocket.java

```diff
@@ -11,255 +11,92 @@
 /**
  * Created by juksilve on 4.6.2015.
  */
-public class BtToRequestSocket extends Thread implements StreamCopyingThread.CopyThreadCallback {
-
-    BtToRequestSocket that = this;
+public class BtToRequestSocket extends BtToSocketBase{
 
     public interface ReadyForIncoming
     {
         void listeningAndAcceptingNow(int port);
     }
-    private ReadyForIncoming readyCallback;
-    private BtSocketDisconnectedCallBack mHandler;
-
-    private BluetoothSocket mmSocket = null;
-    private Socket localHostSocket = null;
+    private final ReadyForIncoming readyCallback;
     private ServerSocket srvSocket = null;
 
-    private InputStream mmInStream = null;
-    private OutputStream mmOutStream = null;
-    private InputStream LocalInputStream = null;
-    private OutputStream LocalOutputStream = null;
-
-    private String mPeerId = "";
-    private String mPeerName = "";
-    private String mPeerAddress = "";
-
-    private StreamCopyingThread SendingThread = null;
-    private StreamCopyingThread ReceivingThread = null;
-
-    private int mHTTPPort = 0;
-
-    final String TAG = "--BtCon-CLIENT-Socket";
-
-    boolean mStopped = false;
-
-    public BtToRequestSocket(BluetoothSocket socket, BtSocketDisconnectedCallBack handler,ReadyForIncoming callback) {
-        print_debug("Creating BtConnectedRequestSocket");
-        mHandler = handler;
-        mmSocket = socket;
+    public BtToRequestSocket(BluetoothSocket socket, BtSocketDisconnectedCallBack handler,ReadyForIncoming callback)  throws IOException {
+        super(socket,handler);
+        print_debug("BtToRequestSocket", "Creating BtConnectedRequestSocket");
         readyCallback = callback;
-
-        InputStream tmpIn = null;
-        OutputStream tmpOut = null;
-
-        // Get the BluetoothSocket input and output streams
-        try {
-            print_debug("Get BT input Streams");
-            tmpIn = mmSocket.getInputStream();
-            tmpOut = mmSocket.getOutputStream();
-        } catch (Exception e) {
-            mStopped = true;
-            print_debug("Creating temp sockets  failed: " + e.toString());
-            mHandler.Disconnected(that,"Creating temp sockets  failed");
-        }
-        mmInStream = tmpIn;
-        mmOutStream = tmpOut;
-    }
-
-    public void setPort(int port) {
-        mHTTPPort = port;
     }
 
     public void run() {
+
         try {
-            print_debug("start Socket with port: " + mHTTPPort);
-            srvSocket = new ServerSocket(mHTTPPort);
-            mHTTPPort = srvSocket.getLocalPort();
-            print_debug("mHTTPPort  set to : " + mHTTPPort);
+            srvSocket = new ServerSocket(0);
+            print_debug("BtToRequestSocket", "mHTTPPort  set to : " + srvSocket.getLocalPort());
         } catch (Exception e) {
```

Comment:
We must not ever catch Exception. This is pretty much always wrong (in the same sense that using GOTOs is always wrong, except, in the few cases where it isn't and this isn't one of them). In this case the obvious problem is that run is on its own thread so any exceptions will be eaten by the thread if they aren't caught.

Except Java has a specific feature to handle this, UncaughtExceptionHandler. This is a method on all Thread objects that lets you register a handler who will be called if the thread exits with an exception. That handler should then repeat the exception on the main thread. Yes, there are other mechanisms like the default unhandled exception handler but it's better hygiene to set handlers as low in the chain as possible.

The idea is that this catch should only catch exceptions we are expecting, like ioExceptions. But by catching the root Exception object we also catch errors that are deadly serious and we must not catch, like RuntimeException and all of its children that include fun things like null pointer exceptions and indexing errors. The kind of exceptions that need to cause the program to fail because something has gone horribly wrong.

So please, use UncaughtExceptionHandler on all of your thread objects in all of your classes and use a generic handler that just re-throws anything it gets onto the main thread. And please only catch here specific exceptions you should be looking for like ioExceptions.